### ISSUE TITLE
optional-gui capability

### ISSUE THREAD

**Issue by grammoboy2** [M0JXD/non-sequencer-fork#3]
Not sure how useful it is to add optional-gui capability. JML didn't needed it and I think I see why, especially for non-mixer. It might be good to have a build option to turn it off and / or to let the NSM client restore the gui state itself after session reloading, which is also mentioned in the NSM API: 

"It is the responsibility of the client to remember the visibility state of its GUI across session loads. " 

http://non.tuxfamily.org/nsm/API.html#n:1.2.4.1. 

cc @Stazed

**Comment by M0JXD** [comment]
> let the NSM client restore the gui state itself after session reloading, which is also mentioned in the NSM API.

I think this already works correctly in my branch. From my rudimentary understanding, all NSM programs with optional GUI, if under NSM, should start hidden (and send the message to say it is). At which point the SM can show it if desired at start-up. This is what I've implemented, and should be the same as Stazed's XT forks. So the feature can be ignored almost entirely if desired.

Regarding whether it's needed, I've found it quite useful as I can have a pattern/sequence being output to Carla, my recorded tracks from timeline playing as well and then only have Non-Mixer shown to Mix without a cluttered Taskbar of things I don't need at that time. Everything having show/hide GUI really enhances the feeling of modularity for me.

I'll add a build option when I merge to main though because afterall it's a new feature that not one everyone necessarily wants. For now that's the reason I've kept it in a separate branch.

**Comment by grammoboy2** [comment]
Hm yeah, I don't see that new API as legitimate, as not approved by the original author and just decided by some people in the background.

```let the NSM client restore the gui state itself after session reloading, which is also mentioned in the NSM API.``` 

What is meant by this, is that if you have the optional-gui of the NSM client shown at the moment of saving and closing the session, it starts shown when you re open your session. Or vice versa, if hidden it starts hidden. If you want to startup all hidden everytime, your NSM gui, could just hide all clients before closing the session.

**Comment by M0JXD** [comment]
I've seen about the Non/NewSM stuff and it's hard to know which to follow. I have the utmost respect for Jonathon, his work on all the Non suite was incredible and I do agree with his a lot of his rational (at least that which can still be found on the internet) behind the technical decisions in the project, but given the one that's actively maintained is the "new" one that's what I've tried to follow, for better or worse. They seem mostly cross compatible anyways.

Here's what's said about [the new API](https://new-session-manager.jackaudio.org/api/index.html#_show_optional_gui):

> If the client has specified the optional-gui capability, then it MUST send this message whenever the state of visibility of the optional GUI has changed. It also MUST send this message after its announce message to indicate the initial visibility state of the optional GUI.
The client SHOULD always start hidden, if not saved as visible. That implies the first load, after adding to the session, SHOULD always be hidden.
It is the responsibility of the client to remember the visibility state of its GUI across session loads.

So yes, you're right regarding remembering the GUI state after the first launch, however Raysession can be set to save the GUI states and it looks like Non-XT (which I referenced heavily to implement this feature) seems to ignore this rule too, relying on the interaction being the other way around.

Also I've just checked and Qtractor even has the optional GUI for NSM, another which could be argued does not require it.

**Comment by grammoboy2** [comment]
> however Raysession can be set to save the GUI states

This is not following the original API and it is also not 100% reliable IIRC. 

> Also I've just checked and Qtractor even has the optional GUI for NSM, another which could be argued does not require it. 

:optional-gui: can be added to a NSM client, so if you like to have it, it's your choice of course.

**Comment by Stazed** [comment]
> It is the responsibility of the client to remember the visibility state of its GUI across session loads. 
 
This is ambiguous, i.e. does "responsibility" mean required? For the clients I maintain, I do not wish to implement saving the gui state option, and always load hidden. Is there anything in the API, original or new, that requires the client to save the gui state. Let me know if you find something that says it is...

**Comment by grammoboy2** [comment]
> The client SHOULD always start hidden, if not saved as visible. That implies the first load, after adding to the session, SHOULD always be hidden.

If I launch a application, I expect it to show up. I think the problem is that some NSM clients doesn't restore their gui state between sessions and so a lot of applications smash into your face, which is especially not convenient when working with a traditional Window Manager. 

Like I said, a NSM gui could have a setting to hide all applications before closing, if they all restore their state, you open a session with all NSM clients hidden. 

It could also be convenient if you launch multiple NSM clients at once, I'm able to do that, but I never use it. A session template (starting with NSM clients hidden) might make more sense when you want to launch multiple applications at once. 

There might be (future) environments where a NSM client can't be hidden? I think NSM should be made for all kind of environments in mind. 

A plugin host can mandate a plugin what to do, this is a setting and in Ardour the default is to show the plugin gui at launch. 

So, no I'm not convinced. I think it's a quick fix, I doubt it is the right way to fix it.

**Comment by grammoboy2** [comment]
> > It is the responsibility of the client to remember the visibility state of its GUI across session loads.
> 
> This is ambiguous, i.e. does "responsibility" mean required? For the clients I maintain, I do not wish to implement saving the gui state option, and always load hidden. Is there anything in the API, original or new, that requires the client to save the gui state. Let me know if you find something that says it is...

See my post above. What if the NSM gui was able to send 'hide' to all clients before closing the session? 

What you could do for now is to make this a easy to configure / build option maybe. E.g. restore the gui state by the client itself or always start hidden. Let the user decide?

**Comment by M0JXD** [comment]
@grammoboy2  - quick question for clarification: Do you use the original Non-Session-Manager?

What I will do is (as discussed) ensure a build option is present to disable optional GUI completely, which should be compliant with original NSM and work with your requirements.

> I do not wish to implement saving the gui state option, and always load hidden.

Me neither to be honest. Ideally I didn't want to have to save the window sizes either, it would be nice to just send them and receive them from the session manager as oppose to the expectation of polluting audio projects with windowing data. At the moment I've saved them globally for testing but I'll add the save per project soon. That way if a session is transferred to a different computer new windowing data can be handled for all clients by the manager.

**Comment by grammoboy2** [comment]
> Ideally I didn't want to have to save the window sizes either,

I think JML would say that this is the task of your window manager.

**Comment by Stazed** [comment]
> I think JML would say that this is the task of your window manager.

Well, I think it is unethical for you to be pretending to be the voice of JML.
And need I remind you of the GPL :

> WITHOUT  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or  FITNESS FOR A PARTICULAR PURPOSE

In short, I will do whatever I want with my implementation and if you don't like it, don't use it.

**Comment by M0JXD** [comment]
> > Ideally I didn't want to have to save the window sizes either,
> 
> I think JML would say that this is the task of your window manager.

I agree (not whether or not that's JML's opinion but the Window manager bit), but if you want different window layouts for different projects the window manager will not be able to differentiate and all saved sizes will be global

He does discuss it briefly [here](https://non.tuxfamily.org/session-manager/doc/MANUAL.html) : 

> NSM manages clients together in a session. That's it. NSM doesn't know or care what Window Manager or audio subsystem those clients use--nor should it. Specific clients must be written to persist these environmental factors, and added to sessions when required.

So I think the solution is an NSM client with the purpose of saving window sizes and positions of GUI clients in a session, similar to what jackpatch does. 

Edit: Maybe this hypothetical client should capture the hidden/shown states too.

**Comment by grammoboy2** [comment]
> Well, I think it is unethical for you to be pretending to be the voice of JML. And need I remind you of the GPL :
> 
> > WITHOUT  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or  FITNESS FOR A PARTICULAR PURPOSE
> 
> In short, I will do whatever I want with my implementation and if you don't like it, don't use it.

Sorry, I do miss both points. If you read well, I say "I **think** that he would say ". What's wrong about that? 

And I don't see how GPL has to do with this. It's about a API. Sure you don't have to follow it, that in general works great for session management (not)...

**Comment by grammoboy2** [comment]
> So I think the solution is an NSM client with the purpose of saving window sizes and positions of GUI clients in a session, similar to what jackpatch does.

I like dynamic window managers to do this for you. In fact most window manager can be configured to allow some 'tiling behavior' I think?  A NSM client who does this is a interesting idea, but I would let it rest for a while probably and see if your preferences for a WM will change over time. My 0.02$.

**Comment by M0JXD** [comment]
Trust me, I'm in love with Mint Cinnamon 🤩
Yes it can tile (I use an extension that improves it too) but it's not the same as a tiling WM.
And even so, not everyone likes to use tiling WMs, an option for those users would be good. Probably something to discuss with the NSM maintaners.

**Comment by M0JXD** [comment]
I'm thinking this could be implemented with something that wraps wmctrl, which is usually installed and should work for most window managers.

The tricky part would be getting the list of clients in a NSM session so it doesn't try to save windows outside the session. I don't know whether it'd worth adding to the API to request a list of apps and what GUI capabilities they claim to have from the SM, and then pass back the GUI states for the SM to send out. Once all shown windows are open wmctrl kicks in to resize and place them.

The other option is using broadcast messages to request that GUI clients send broadcasts back with their name and GUI capabilities (no-gui, always-visible or optional-gui).

That way developers of NSM applications needn't worry about saving aspects of the GUI (and worrying about how that works across different window managers. They can always start hidden with a default size.

I'd like to know the thoughts of some of the prominent NSM developers regarding this but given it's Christmas I'll ping them in the new year :)

**Comment by grammoboy2** [comment]
About storing data, non-sequencer is storing a file in the session directory. Made by JML, but I think it's better that NSM clients make their own project folder and leave the session directory for stuff that concerns the session as a whole, like session.nsm, jackpatch file, session notes file etc. 

Note you're dealing with X11 and especially Wayland nowadays. 

I think :broadcast: is not what you're looking for, then all NSM clients needs to support this. It's hard enough already to get NSM support into clients. 

There is no reliable way to know if a session is loaded I think. Yes NSM clients get a message, but not with a error code, so you don't know if the session is loaded successful, e.g. if all NSM clients you need loaded successful. Non-daw relies on broadcast therefor I think.  

Passing back to the daemon and intervene how to handle gui states, doesn't sound like a good plan to me tbh. 

What information do you need from the daemon? In theory that could be delivered I think, if it's worth it, is another question. xdotool etc also use PIDS right? How does that work for Wayland? Sending pids to other tools to mess with, hm ...

https://manpages.opensuse.org/Tumbleweed/wlrctl/wlrctl.1.en.html

**Comment by grammoboy2** [comment]
> What information do you need from the daemon? In theory that could be delivered I think, if it's worth it, is another question. xdotool etc also use PIDS right? How does that work for Wayland? Sending pids to other tools to mess with, hm ...
> 
> https://manpages.opensuse.org/Tumbleweed/wlrctl/wlrctl.1.en.html

Did you try raysession? It has it build into it's daemon it seems and uses pids: 
https://github.com/Houston4444/RaySession/blob/e2714b8c3bf3b443048d8f32614fe652cabe798d/src/daemon/desktops_memory.py#L27 

I think this could work as a NSM client, but then the daemon (it's dev) must be willing to send the pids, which I personally find questionable. I wouldn't build it into the daemon. A third option is to do it in the NSM GUI, which then also needs the pids, which raises the same design questions.  

I would say try Raysession and see how it works for you and how much you really need desktop memory. I personally think you add just another possible point of failure to your session and that your window manager is much better capable of doing this sort of things.

**Comment by M0JXD** [comment]
> Did you try raysession? It has it build into it's daemon it seems and uses pids:

I actually wrote out a draft for your previous response last night to proof read before posting today. But now you've posted this new reply it's half redundant lol.
I'll still post it as it covers everything I wanted to discuss, but in short, I do use RaySession, however as a developer I want to know that the behaviour works across any NSM server.

Here it is: 

> About storing data, non-sequencer is storing a file in the session directory. Made by JML, but I think it's better that NSM clients make their own project folder and leave the session directory for stuff that concerns the session as a whole, like session.nsm, jackpatch file, session notes file etc.

A single .non file is the _only_ file needed/created by a Non-Sequencer instance, regardless of the amount patterns/phrases/sequences. Hence it doesn't create (perhaps wrongly) it's own directory when under session management. However .non files are actually a variation on [SMF2 MIDI files](https://non-sequencer.tuxfamily.org/MANUAL.html#n:3.2.), it would be very wrong to put windowing data in there.

> What information do you need from the daemon? In theory that could be delivered I think, if it's worth it, is another question. xdotool etc also use PIDS right? How does that work for Wayland? Sending pids to other tools to mess with, hm 

wmctrl is not the only tool for this of course, but it was probably the most accepted prior to Wayland (it's actually used by RaySession). This is another case of Wayland being not yet fully cooked because it lacks a full featured, generally accepted replacement. Xdotool is not installed by default on my system and it is not compatible with Wayland anyways.

The only information that's "needed" is some way to discover what applications are actually in the NSM session out of all possible running applications, otherwise it would try and capture apps outside the session like your web browser and would also be unaware of "hidden" applications. 

As said, I'd like to see what other NSM developers have to say on this. After all, RaySession can save UI states and force them to be as previously saved, which is the opposite to what is specified in NewSM API documentation. Both mine and Stazed's forks rely on this incorrect behaviour, and I think I speak for both of us in saying we prefer it as well. As such, a new client for this task would only have use under the (new/old) nsm daemon, but that is practically never used outside of Agordejo.

Session management is complicated. My "ideal" would be that a modular, session managed audio project does not need to save anything aside from it's relevant audio data and localised data specific to the task it handles (e.g. snapping modes, zoom, rulers/grids, what tracks are enabled etc.). 

Window management is something that can be different per project, but also affects how all these modular applications work together on a display system, much akin to how JACK patching controls how they work together in the audio system. As such it I think it's management should be outsourced to something external. As a developer, I merely provide a way for something external to set the visibility, size (if this is even needed, the WM can do this. I think you would have to tell the WM the desired positions/virtual desktops/screens anyways) and provide a suitable startup condition under session management (hidden in this case).

Given I don't want to save such windowing data in my projects as a client, the next question is does such a responsibility fall on the session manager or a client of the session manager? The initial vision seems that the session manager stays true to it's name. It saves all the apps data under one place and has basic control of it's clients, that's it. It should not be care for controlling JACK, the WM, or whatever. If you need that, you have a client that it can start, stop and pass generic messages to.

However my favourite session manager is RaySession. It's aware of JACK, UI states, and virtual desktops. It's the opposite of this ethos and yet I highly prefer it's feature-set. 
A discussion amongst prominent users of NSM is probably required for a way forward. What is there, a couple hundred apps with NSM support? Done by maybe 50 or so developers? There's only two actively developed NSM servers anyways!

I'm still very new to the Linux Audio scene and audio development (arguably to programming in general) but NSM is something I'm keen to use. I think the concept of modularity in a virtual studio is great, it's analogous to how a real world studio is. But it needs standards everyone can agree on. Imagine the music world if the plugs, pin-outs and protocol for MIDI was not open and standardised.

**Comment by grammoboy2** [comment]
> The only information that's "needed" is some way to discover what applications are actually in the NSM session out of all possible running applications, otherwise it would try and capture apps outside the session like your web browser and would also be unaware of "hidden" applications.

This is possible. But how do you distinguish between two versions of let's say zynaddsubfx though? I've little experience with WM related stuff. If you've a prototype which ideally (also) works in Wayland, I might see if it could work, all though I'm not going to use this myself. 


> As said, I'd like to see what other NSM developers have to say on this. After all, RaySession can save UI states and force them to be as previously saved, which is the opposite to what is specified in NewSM API documentation. Both mine and Stazed's forks rely on this incorrect behaviour, and I think I speak for both of us in saying we prefer it as well. As such, a new client for this task would only have use under the (new/old) nsm daemon, but that is practically never used outside of Agordejo.

This can be done in a different way I think, which is also backwards compatible API wise. I've already tested this. The question is it worth it given that optional-gui is not a essential feature and if a NSM client doesn't act perfectly, it's easy to ignore. Currently it might do more harm then good to introduce this. So I'll sit on it for some more time I think.  

> Given I don't want to save such windowing data in my projects as a client, the next question is does such a responsibility fall on the session manager or a client of the session manager? The initial vision seems that the session manager stays true to it's name. It saves all the apps data under one place and has basic control of it's clients, that's it. It should not be care for controlling JACK, the WM, or whatever. If you need that, you have a client that it can start, stop and pass generic messages to.

Correct. This is how you control a very complex system with many components in a complex Linux/OpenSource/GPL environment. It also is the reason why NSM survived among other session protocols imho. 

> However my favourite session manager is RaySession. It's aware of JACK, UI states, and virtual desktops. It's the opposite of this ethos and yet I highly prefer it's feature-set.
> A discussion amongst prominent users of NSM is probably required for a way forward. What is there, a couple hundred apps with NSM support? Done by maybe 50 or so developers? There's only two actively developed NSM servers anyways!

So why don't you use raysession? 

Way going forward? What do you mean exactly? 
Probably not fixing building non-sequencer (thx!), fixing some issues in JACK, improving JACK/Pipewire for session management, adding JACK and NSM support to SurgeXT or Vital is it? ;)

**Comment by M0JXD** [comment]
> Way going forward? What do you mean exactly?

I mean for Window Management in the context of NSM, given the two primary session managers (RaySession and Agordejo) have different behaviour and expectations for how it should be achieved.

If I'm developing an NSM application, I want to provide support for whatever session managers users might use. I do use RaySession and hence that's what I've built my feature around, but if someone wants to use my mildly adapted version of Non Sequencer under Agordejo (because we all have preferences and open source tools are for all the community to use and enjoy) I would prefer it to be supported and behave as they expect. Agordejo expects the client to save it's UI preferences.

Telling Non-Sequencer-Fork (I might need a better name if I keep working on it) users that the only _fully_ supported session manager is RaySession would prevent it's adoption and use. If I start having to adapt my client to behave correctly per session manager (offloading saving the state when under RaySession, and saving it myself when under nsmd/Agordejo) that begins to defeat the point of the defined NSM protocol.

**Comment by grammoboy2** [comment]
You don't add support for a session manager implementation but for a session management protocol or API, the NSM API. It says that the NSM client is responsible for restoring it's gui state, which means either hidden or shown (visibility state). 

But are you talking about visibility state only (hidden or shown) or actually about something more?

**Comment by M0JXD** [comment]
> But are you talking about visibility state only (hidden or shown) or actually about something more?

Well, I think something more. [See the picture in my feature request for Non-Mixer-XT](https://github.com/Stazed/non-mixer-xt/issues/47#issuecomment-2558634283). This is GUI layout I've found myself liking. When using a session manager I want all my windows to open the same way as I last had it, with the same sizes, positions, visibility and on the same virtual desktops. I also think my client shouldn't have to save and set all that information. RaySession saves some of this information. Agordejo does not save this information at all. The API is the same but the definitions/implementation is different. If I want this behaviour to be consistent across session managers, I'll have to adapt the client for each of them.

Let's get back on track. You don't want the optional GUI in Non-Sequencer, I've said I do. As a compromise I've said I'll provide a build option for it when I merge the branches. 

You've then complained that I'm using the wrong NSM API/implementation.

As I see it, there's three options:
+ The original Non version, which no one is maintaining. You've implied this is the one you want.
+ The New version as being maintained by the Nils/the JACK Audio team.
+ The implicit version that RaySession offers, which breaks the definitions of the New version by saving GUI aspects.
 
My conclusion is that:

1) I most prefer RaySession's version, and have implemented the GUI behaviour to be correct for that.
2) I won't support the original Non version because it's completely unmaintained, not even for bugfixes. Luckily it's mostly cross-compatible.
3) Ideally there would be a way for the client to have the "RaySession experience" under the "official" new version, be it by changing the expectations of the official implementation, or developing a client (akin to jackpatch) for it, because I dislike the idea of saving windowing data with my project files.

After this point I gave some brief thoughts as to how such a client _might_ work but have said that before anything the thoughts of the prominent NSM application developers and the official maintainers is important to obtain - but it is Christmas and I don't want to bother them right now. We've then speculated the merits of such approaches and also the difficulties of ensuring it's actions are consistent between X11 and Wayland.

You've also suggested that that this won't be needed because you like to use a dynamic WM which can take care of such things. However mine can't take of care of this. You've then implied I should change my WM preferences to something with the same capabilities. It's unreasonable to expect everyone to have the same preferences as yourself.

Regardless, I've expressed interest in making sure that my client should behave as expected regardless of WM or SM. Changing my personal choice of WM fixes nothing in this regard.

Can you see the difficulty I face? Without developing a specialised GUI client or changing the API expectations, if I:

1) implement it the way you want, it will not work under RaySession as I would like.
2) implement it the way the NewSM maintainers want it, it will not work the way either of us would like.

With both of these solutions I'm saving window data in the project which I've already expressed distaste for.

Let's wait on some new opinions in the new year. Anyways, the problem can be avoided entirely for you by building the branch without the optional GUI feature, and using the build option I'll provide when the time comes.

**Comment by grammoboy2** [comment]
> Well, I think something more. [See the picture in my feature request for Non-Mixer-XT](https://github.com/Stazed/non-mixer-xt/issues/47#issuecomment-2558634283). This is GUI layout I've found myself liking. When using a session manager I want all my windows to open the same way as I last had it, with the same sizes, positions, visibility and on the same virtual desktops. I also think my client shouldn't have to save and set all that information. RaySession saves some of this information. Agordejo does not save this information at all. The API is the same but the definitions/implementation is different. If I want this behaviour to be consistent across session managers, I'll have to adapt the client for each of them.

It's not your task as NSM client developer to fix this for your users. Simply put, it's not your problem. One of the available session managers provides a extra feature, namely the use of wmctrl to save desktop information. That can be done without having the NSM API for clients being involved. So if your users wants this feature, they should use raysession or ask the Agordejo dev to provide it. 

Raysession wants to provide a lot for the users, makes application specific exceptions and I understand that it can be nice for users. I do understand that people like the integrated patchbay and I personally like the idea of client templates a lot. It's a remarkable piece of hard work, which is not easy to make it all work. The other side of the coin is that it needs 22225 lines of python code to provide this, this is without patchbay, which makes it 60000 lines of python/qt code (and with all the submodules 95000 lines of code). This needs to be maintained and this has influence on reliability, portability and future sustainability. Compare this with non-session-manager/ nsmd, it has 5000 lines of C++ or so I think. 

You've suggested to write or possibly experiment with a NSM client which saves desktop data. I did ask you how you would do that, how you would distinguish between two versions of Zynaddsubfx in the session. You didn't provide further information yet. I also said that I think this is in theory possible, especially if you don't need pids. So this still can be worked out. 

A third option is to do it in the NSM GUI I think. 

From these options I would prefer to look at what is possible with a NSM client as a thought or actual experiment, but I prefer the fourth option myself at the end: leave it to the WM. But that's just a personal opinion, I didn't say you must use another WM. You should ask yourself whether you really need this feature, like with all ideas and feature requests. Sometimes perfection is the evil of good. 

Exchanging preferences is not how a API works. If I provide you such, then I can violate a bit here, so then we're equal. A API should be general applicable and a API does not cover every usecase, preference or exception and it shouldn't, cause it will become a mess quite soon. Personally I try to avoid such developers or fix it in my own setup, cause it's a endless constantly repeating discussion. 

Let's end with a quote of the day for me: 

"The feeling that tech's promise is to make everything easier yet its constant failures and inconsistencies and growing complexity are making everything much, much harder.

Or to put it another way, spending far more time trying to make things work than their working will ever save us."

https://hachyderm.io/@robpike/113715650138654284

**Comment by grammoboy2** [comment]
As far as I can see you need pids. This also should work on wayland. But note, does raysession support wayland? Wmctrl does not afaik.  At least raysession now also needs to check for X11 vs Wayland afaik. Fltk based applications doesn't send pids though: 
https://github.com/Houston4444/RaySession/blob/master/src/daemon/desktops_memory.py#L109 

So there is a exception hardcoded for it, using explicit application names (non-mixer, not non-mixer-xt). 

The daemon has the most reliable access to pids. So sending pids to other tools is also a design choice which has to be made. 

One can choose to implement this desktop related stuff, but I think it should also pretty clear that there are good reasons to avoid such things in a session manager. And we're just talking about Linux only here, but in theory NSM could work crossplatform.

**Comment by M0JXD** [comment]
> You've suggested to write or possibly experiment with a NSM client which saves desktop data. I did ask you how you would do that, how you would distinguish between two versions of Zynaddsubfx in the session. You didn't provide further information yet. I also said that I think this is in theory possible, especially if you don't need pids. So this still can be worked out.

I didn't provide more information because I don't really know! It's a very complicated problem.

I think your right about using PIDs to know what to save, but that only solves half the problem. When restoring windowing data for the next time the session is started all PIDs will be probably be different anyways. Like your example of two instances of ZynAddSubFX, I'm not sure how a such a client is supposed to know which instance to apply what windowing data too. I can't think of a way around this without some context coming from the session manager.

There's a request for such a feature in [RaySession](https://github.com/Houston4444/RaySession/issues/105) where Houston4444 discusses the difficulties from the server developers perspective. RaySession's desktop remembering features are not guaranteed for Wayland.

**Comment by grammoboy2** [comment]
> Like your example of two instances of ZynAddSubFX, I'm not sure how a such a client is supposed to know which instance to apply what windowing data too.

With pids you can match them. 

@Stazed What do you have against the API rule to let clients restore their own state? I know this a very explicit design choice of JML. 

Is it just that you want your application always to be hidden. Or don't you want to store the visibility state (hidden or shown) and if so why not?

Seq192 is a example which stores it's own visibility state across sessions.

**Comment by Houston4444** [comment]
Hi !

"Save/restore GUI state" RaySession option has been done because almost programs at that time were not doing the job their self, and I was perfectly understanding that it can be annoying to develop. I still think it is better this way (job is done by session manager => less work for a NSM implementation). But of course, the program can do the job itself and has not to care about this RaySession option. If something is wrong with this option with a program saving its GUI visibility, then a bug report to RaySession is welcome.

**Comment by M0JXD** [comment]
No problems with RaySession, it's perfect!

My difficulty is making sure that the optional-gui behaviour is consistent between session managers because RaySession and Agordejo are different. I think I'll just specify that RaySession is the best supported manager for optional-gui feature, and there's a build option to disable it completely anyways. If Agordejo/NonSM users request changes that's something I'll look in to as and when.

Thanks so much for your input.

**Comment by grammoboy2** [comment]
Knowing JML, most of the time there are good reasons why he did something is my experience. Here I found a short comment: 
https://www.mail-archive.com/non@lists.tuxfamily.org/msg00455.html 

I looked at it before:
https://github.com/Houston4444/RaySession/issues/168 

And concluded that it's not 100% reliable if you overwrite this the way raysession does.  

Letting a NSM client hide at launch as matter of API, feels very unsatisfying for me. There is no choice to show it at launch. Showing the gui by default is most of the time the most sane thing to do. See also Ardour, which shows the Gui of a plugin by default. Meh.

**Comment by grammoboy2** [comment]
Nah after much doubts and tests I think I actually have seen the light and agree with JML here. 

If you let the the NSM save it's own visibility state, then the project folder is a client template in itself and I realize I already use this in my own setup. 

If you let the NSM client save it's own visibility state, then the session will exactly be restored the way you left it. 

If a client resumes, it opens it's project folder and restores it's visibility state of it's last save and the daemon doesn't need to be involved for it. 

It is reliable this way, no conflicts between client and daemon and accomplished with a simple API (I had a idea which had much more complexity involved tbh). 

I think the daemon needs a hide all NSM clients method. A NSM GUI could be set to first hide all clients before closing probably. 

NSM clients should just show up when they're added to the session. 

My conclusion for now.

**Comment by M0JXD** [comment]
Yes, I see your points. Given we've established it would be really hard to ensure either a server or dedicated client can reliably restore sizes and positions of windows in a NSM session across different desktop environments and windowing systems it's probably for the best that it tries to save it itself. For now, I've made it so the window sizing is saved as a global config value. This is only in the hide/show GUI builds. With that turned off it should always start at default size. I'm not sure whether I'll implement this per project as I see little reason for it and as stated it pollutes audio projects with windowing data.

> NSM clients should just show up when they're added to the session.

Given NewSM and RaySession expects clients to default to hidden when added I've done that instead. Of course without optional GUI it must start shown so you can use that - you've said you have little you use for optional GUI anyways.

I think I can close this now.

**Comment by grammoboy2** [comment]
> Given NewSM and RaySession expects clients to default to hidden when added I've done that instead. Of course without optional GUI it must start shown so you can use that - you've said you have little you use for optional GUI anyways.

If you make this a config option or a build option, then you give your user a way to add a client with gui shown. 

I'm pretty much a fan of suckless.org approach, where you can easily set these things in a config.h file. 
https://dwm.suckless.org/customisation/ 

This approach works if the application is easy and fast to build (like suckless tools should).

**Comment by grammoboy2** [comment]
> Given NewSM and RaySession expects clients to default to hidden when added I've done that instead. Of course without optional GUI it must start shown so you can use that - you've said you have little you use for optional GUI anyways.

If you make this a config option or a build option, then you give your user a way to add a client with gui shown. 

Maybe this should even be a API 'should' rule to provide such configuration option, as a way to always be able to change this setting, as it feels conceptually wrong for me to hide by default, but especially if it's the only option. 

I'm pretty much a fan of suckless.org approach by the way, where you can easily set these things in a config.h file. 
https://dwm.suckless.org/customisation/ 

This approach works if the application is easy and fast to build (like suckless tools should).

**Comment by M0JXD** [comment]
As oppose to another build option I've instead opted for config approach where the HIDEGUI build checks for the existence of a file called "force_show" in the .non folder at startup. Hence if package maintainers start packaging this code it remains configurable for end users without requiring they compile it.

If this file exists (it's contents do not matter) it will force show the GUI at startup. You will have to make the file yourself but I do not think this as a major issue. I can't seem to get the RaySession and Agordejo to understand this behaviour (cycling the GUI command in these a couple of times sorts it though) despite sending the "is shown" NSM message, maybe because it is incorrect behaviour as per the NewSM standard. Because it's not standard I won't investigate this further (it behaves correctly without the override). This is _possibly_ a problem with the session managers. You're welcome to investigate further as to why.

**Comment by grammoboy2** [comment]
> This is _possibly_ a problem with the session managers. You're welcome to investigate further as to why. 

Ha, no thank you... what do you think?  I've much better things to do, then dealing with incompatibility problems caused by your in this case. It's makes NSM frustrating as hell. 

Anyway, so for what it's worth, I think it would acceptable to share pids of NSM clients with other tools as exception for use with tools like wmctrl and as setting which is disabled by default, if someone can show such tool is reliable and doesn't affect the overall reliability of the system as a whole, including false expectations users might get. I think you don't want a session management system where all kind of tools are able to mess with pids. Also I wouldn't accept literal coding exceptions for specific applications in either the daemon or NSM GUI, but that might be acceptable in a external maintained NSM client. 
```
Save the window size (currently, this is saved for all instances under .non/window, saving per project is not implemented and will not be until someone either provides a good reason to or submits the code.)
```
If raysession is saving this for you, why do you need to save it as well? Is this common for Desktop applications to do this kind of stuff, isn't that seen as the task of the WM in general?

**Comment by M0JXD** [comment]
> Ha, no thank you... what do you think? I've much better things to do, then dealing with incompatibility problems caused by your in this case. It's makes NSM frustrating as hell.

This is why I was moaning about the differences between the session managers, ensuring compatibility is proving a bit difficult. I did ask which session-manager you are using? Is it the original Non?

I actually have looked a bit more and I reckon it needs the [command_open](https://github.com/M0JXD/non-sequencer-fork/blob/280ea96b665075319a968c4ce775d287d6b51c6c/src/NSM.C) changed to do similar to what Non-Timeline/Mixers does with the [say_hello method](https://github.com/Stazed/non-mixer-xt/blob/main/mixer/src/Mixer.C), and send the "is_shown/hidden command" there. The [NSM API](https://new-session-manager.jackaudio.org/api/index.html#_optional_gui) says it _should_ be reported after the announce but I think it actually needs to reported as a response to the open command. JML didn't refactor Non-Sequencer's open command with the "say_hello" like Timeline and Mixer have been so the NSM stuff is sort of mixed in with the general opening stuff.

> If raysession is saving this for you, why do you need to save it as well? Is this common for Desktop applications to do this kind of stuff, isn't that seen as the task of the WM in general?

Sorry for misleading you, this is something RaySession doesn't do. I mentioned the feature request that was made for it [here](https://github.com/M0JXD/non-sequencer-fork/issues/3#issuecomment-2563749630). Houston4444 reckons a separate client would never be able to work perfectly, and it's also really hard to ensure this feature would work correctly as part of the session manager too. I have no reason to disagree, so saving the size with the client is the only option really. Non-XT does this per-project but I think it's better as a global config as to not have windowing data stored with audio projects. What RaySession does do is remember GUI shown/hidden states and the virtual desktops.

**Comment by grammoboy2** [comment]
say_hello is not part of the NSM API, it's part of Non-Daw which uses the NSM capability :broadcast:. http://www.non.tuxfamily.org/nsm/API.html

You're aware that WM's can make applications appear on a specified virtual desktop number? Some can set to floating vs tiling, it's probably possible to set the size as well on some WM's. So then you've a global setting. 

I looked a bit closer to wmctl and it seems to be a hack in itself, might be useful in some cases, but seems  a fragile system to me, where you've to dump all data. It depends on x11 (wayland might provide a alternative?), so you make your NSM session depend on x11. It seems to depend on the resolution setting you use, your monitor setup (multiple monitors). So you make your NSM session depend on all of that. 

This is also partly true if you store sizing data in the project folder of course, or must be scaled relatively to the screen resolution you use. 

If a NSM client reports to be ready, wmctrl can't always find it's window PID yet, a delay of 200-400ms or so seems to 'fix' this. The optional-gui state is reported a bit later, so that doesn't seems to need a delay if the wmctrl action is based on that.  
It seems reasonable to resize the application just the first time it shows up. If the NSM client which manages wmctrl dies, you've to restart it, it might do the resizing again. 
You see the 'resizing part' live, e.g. window opens on pos x and wmctrl moves it to a other place a fraction later. 

I'm testing on LXDE and all though it's fun to see it happen, I don't see it as something that would significantly improves workflow. A dynamic tiling WM does a much better job imho. Also, how static are your window positions? At some time you will resize and move windows right? 
For now his system seems way to fragile to me to base NSM related choices on it. 

I stay with my opinion that a WM should be that one that has to do the job. Most WM's are not really made for a modular workflow probably. Maybe not strive for perfection. Modular approach has up and downsides.

**Comment by grammoboy2** [comment]
> I don't see it as something that would significantly improves workflow. A dynamic tiling WM does a much better job imho. Also, how static are your window positions? At some time you will resize and move windows right?

ok, moving a window of a NSM client to a other desktop looks to be a more useful feature to me. But this also can be done by the WM, but not session specific.

**Comment by grammoboy2** [comment]
Ok, even though I don't use such a mainstream (Windows inspired?) WM, which is far from ideal for a modular workflow imho (did I already said that?), I did some tests with wmctrl in LXDE today and came up with the following wmctrl related behavior, which seems to work best for me, without being to disruptive :


On session open:  
- let wmctrl move and resize NSM client related windows to the saved virtual desktop number if it is saved with a different then current desktop number;  
- let wmctrl do nothing if NSM client has capability optional-gui, shows up and is saved with the same desktop number as current;  
- let wmctrl minimize a NSM client if it is saved with the same desktop number as current and doesn't have capability optional-gui  
- do all these steps only once, e.g. at session open or when optional-gui NSM client shows up for first time and is saved with a different desktop number then current; 

On session close:  
- let NSM GUI hide all NSM clients first when closing session. 

(put the wmctrl based client in a 'fake modus' if the detected desktop environment is not supported (wayland atm), e.g. then that client does not try to restore or save data).  


Note that moving to a other desktop number could be done by the WM (global config) and has the advantage that you don't have a visual interruption (first show up here and then moved by wmctrl). 

This tested setup is not a solution for fltk tools without optional-gui capability, as wmctrl can't handle those. 

But really, if you've non-daw as your core system, I recommend to start looking at something like i3 (x11) or sway (wayland) (https://swaywm.org/). Sway provides a way to resize windows with your mouse in a flexible way, while this is cumbersome to do in LXDE. 
You can quite easily switch to Cinnamon for more Desktop based workflow when desired. No, you don't have to. ;)

**Comment by grammoboy2** [comment]
It's possible to get a very small delay which gives a visual interruption like you see (or saw) when launching seq192 or seq32. The wmctrl based NSM client could have different configuration options to set it to minimize windows or resize them and to include or exclude certain desktop environments. 

So my conclusion for now is that it's a hack, but it can be a useful hack in some cases. 

If you use a WM like Cinnamon I can understand that Raysession works nice for you. It doesn't restore Window positions and sizes right? It uses wmctrl only for moving to desktop numbers it seems, looking at the code.

**Comment by M0JXD** [comment]
Hi again,

Sorry for my delayed response! I have been reading your comments and meaning to reply but you know how it's like getting into the new year.

> say_hello is not part of the NSM API, it's part of Non-Daw which uses the NSM capability :broadcast:. http://www.non.tuxfamily.org/nsm/API.html

I think I can probably fix SM showing wrong GUI state behaviour by ensuring it's reported after the open command. If you look in Non-Mixer-XT's Mixer.C you'll see that's where it being reported. I think your right that most of the function is more of a broadcast message to alert "Non-Friends" but the SM seems to require the "is_shown" message be sent at that point, even noted by Stazed that it's required for RaySession:

![image](https://github.com/user-attachments/assets/b085804c-521c-4248-8ac0-5eba59102156)

Regarding the sizing and placement, I'm really not sure anymore. My ideal would be that a session restores exactly as it closes, deployed to the same sizes, positions, desktops, tiles etc.. I wouldn't say my window positions are unchanging, it's more about "picking up where you left off". 

For that to work properly that requires one of:
1) Expect every NSM client implements it's own window saving aspects.
2) Have a specialised client using hackey techniques (that don't work on wayland) as you've outlined.
3) Have it be part of the session manager with the same hackey techniques, but that will at least make finding PIDS for what is actually in the session a bit easier.
4) Expect users to utilise a WM that has such ideology ingrained, like sway as you mentioned.

I think for advanced users, expecting them to make their own scripts with wmctrl/xdotool/ydotool/whatever like you've been playing with is maybe not too unreasonable an expectation. It seems pretty hard to make things simpler. I appreciate your findings and insights on this! Thanks.

Have you seen these? Maybe an NSM wrapper around a tool that solves this problem in the general domain could be a better solution.
https://github.com/johannesjo/linux-window-session-manager
https://extensions.gnome.org/extension/4709/another-window-session-manager/ (GNOME only)

**Comment by grammoboy2** [comment]
> but the SM seems to require the "is_shown" message be sent at that point, even noted by Stazed that it's required for RaySession:

Frankly, that doesn't belong in that broadcast function.

The API says the following: 


---
> "If the client has specified the optional-gui capability, then it MUST send this message whenever the state of visibility of the optional GUI has changed. **It also MUST send this message after it's announce message to indicate the initial visibility state of the optional GUI.**
> 
> It is the responsibility of the client to remember the visibility state of its GUI across session loads."

http://non.tuxfamily.org/nsm/API.html#n:1.2.4.1.   

---
> Have it be part of the session manager with the same hackey techniques, but that will at least make finding PIDS for what is actually in the session a bit easier.   

You still need wmctrl to find the window ID's afaik. 
If you use wmctrl, I don't think you win a lot if anything at all by putting that in the daemon in terms of smoothing it,  or I must miss something. You know a fraction earlier which pids you've to search for, but not sure if that's significant enough to notice. Maybe Raysession has a better way or doesn't it do window placement? 

> Regarding the sizing and placement, I'm really not sure anymore.  

It's not easy to fix this in a satisfying manner for all cases probably. 

Tbh I don't really like to dive into this Desktop stuff, as it's not really my problem, but at the end I'm glad I did dive into it, as this discussion comes up regularly of course and it has it's basis in different environments people uses. And when I've to help my younger familymembers around jammin' a bit, a wmctrl tool might come in handy actually. And it has as advantage that you can use it session specific.  
So you screwed up two days of my life, with bringing this up, but at the end it was worth it. ;)

I'm sure wmctrl isn't the only way to do this and it wouldn't surprise me if there are better modular ways to do this indeed. If you want to serve the 'mass' and newbies and want to make all of this really smooth, you're probably moving away from the modular approach.

**Comment by grammoboy2** [comment]
A interesting observation:  if you've a tool which resizes windows for you, it becomes a disadvantage to temporarily hide the gui, as it can't reasonably be restored to the last position imo (or you must save). And restoring to the original position (from session open) is also very questionable if the user wants that, that will be annoying most likely. 

If you want your windows in a specific position, you probably want to keep them visible and use virtual desktops to layout your applications, also on a traditional WM like LXDE. 

When you minimize the application instead, it will gets restored to the same position (LXDE).

**Comment by M0JXD** [comment]
> So you screwed up two days of my life, with bringing this up, but at the end it was worth it. ;)

Glad to here it was somewhat worthwhile :)

The GUI state reporting to the NSM server is now fixed :)
Now if force_show is set it reflects in RaySession and Agordejo. I think the original Non should work too.

If optional-gui doesn't fit someones workflow (I see your point that virtual desktops can render it a bit moot) they can always build without the capability.

I might change this to a run time configuration like the force_show_gui (maybe called no_opt_gui) as it means there would be no requirement for multiple binary packages. Let me know what you think.

**Comment by grammoboy2** [comment]
That you provide this option is already a good thing imo.  In the original API there is another option to force a application to show it's gui (which the fork removed), when the server reports no optional-gui functionality (which it could made capable to fake this with a command line flag), then the nsm client should always shows it's gui.

```
1.2.1.2. Response  
optional-gui:  
	server responds to optional-gui messages--if this capability is not present then clients with optional-guis MUST always keep them visible  
http://non.tuxfamily.org/nsm/API.html#n:1.2.1.
``` 
nsm.h
```
  if ( strstr( nsm_get_session_manager_features ( nsm ),     
                  ":optional-gui:" ) )                                  
             {                                                          
                 nsm_set_show_callback( nsm, cb_nsm_show, 0 );          
                 nsm_set_hide_callback( nsm, cb_nsm_hide, 0 );          
             }  

```

Then you probably need to change the way the application handles quit from the application itself, does it hide or does it stop, depending on whether the server has this functionality or not.  


One other note about the wmctr based application.  The fastest and most simple way to handle this is probably to start searching for the window pid/id whenever the pid of the nsm client is known (could be before announced) and treat applications with optional-gui capability the same as those without. This works in the most simple way if the nsm clients gui's are restored as they're saved, otherwise that wmctrl based application searches for nothing. If that's not the case or the nsm server even could overrule the visual state (not according to nsm api, but raysession can do this) then it complicates things imho.

**Comment by M0JXD** [comment]
> That you provide this option is already a good thing imo. In the original API there is another option to force a application to show it's gui (which the fork removed), when the server reports no optional-gui functionality (which it could made capable to fake this with a command line flag), then the nsm client should always shows it's gui.
> 
> 1.2.1.2. Response  
> optional-gui:  
> 	server responds to optional-gui messages--if this capability is not present then clients with optional-guis MUST always keep them visible  
> http://non.tuxfamily.org/nsm/API.html#n:1.2.1.

This is a request for the maintainer of your chosen session manager.
The new API says this:

![image](https://github.com/user-attachments/assets/c743f020-a72a-4d27-a3a8-9a3606f3768b)

So no changes required right now. If this gets implemented in one of the existing managers (or a completely new NSM program is made with the option) then I'll change it to ensure that this is what happens. Right now there are no session managers to even test this with as New nsmd, ray-daemon and even the old Non nsmd all report optional-gui control. I can't really program this feature without something to test it with! :) 

> One other note about the wmctr based application. The fastest and most simple way to handle this is probably to start searching for the window pid/id whenever the pid of the nsm client is known (could be before announced) and treat applications with optional-gui capability the same as those without. This works in the most simple way if the nsm clients gui's are restored as they're saved, otherwise that wmctrl based application searches for nothing. If that's not the case or the nsm server even could overrule the visual state (not according to nsm api, but raysession can do this) then it complicates things imho.

I think my focus for now is ensuring this fork of Non-Sequencer remains a properly behaved NSM client. If more complex window management becomes a part of one session managers or it gets implemented as a specialised client I'll happily make the changes to ensure Non-Sequencer-Fork respects them and works with them as much as possible. As you said:

> It's not your task as NSM client developer to fix this for your users. Simply put, it's not your problem.

For now I think Non-Sequencer-Fork is about as correct as possible in this regard, as it matches what other clients like Non-XT and Carla are doing.

**Comment by grammoboy2** [comment]
> I did ask which session-manager you are using? Is it the original Non?

Ok, I ignored that on purpose, but I build my own and it's becoming pretty good I think. So, yes it also stores desktop positions now, but it has many other interesting ideas and features I think. Sadly enough I see no way how I could release this any time soon. They created a precedence with their fork. Free software is indeed under attack in this area. I could loose my project the very next day,  they could change the API whenever they wish. Mind you they changed it within a week, without consulting the original community. It's a question of power and some took it from the original creator, made it their own and they can do it again whenever they wish. I guess I've to wait until those people are not active anymore, promise to not touch my project or when new-session-manager is renamed to agordejod or something, but it's hard to see that happen. I think the result of the fork are two death projects and no way forward to be honest, which you could have seen coming of course, but nobody wanted to listen and especially most people just didn't care.

**Comment by M0JXD** [comment]
My personal viewpoint is not to let a minority of bad actors abuse my fun in the free software space. I know enough programming to be able to make whatever (FLOSS) programs I use work the way I need them to (even if the process is not simple or easy!). In such a situation that my changes are not accepted by the original maintainer I'd just use my own forks. If someone forks a project I work on with changes I don't agree on, it doesn't stop me from using/working on my original version. 

What isn't acceptable is purporting that that a forked version is somehow the "official community version" or implying that the original version was inferior adware/spyware riddled software, and using a name that makes it hard to differentiate it from the original (at least the word "fork" in mine makes things pretty clear!). There is very little that's "official" in the free software space anyways, most Linux distros aren't POSIX compliant. Also, blocking people from using the being able to use the community spaces (mailing lists etc.) was poor form too.

Anyone who's spent any time with NSM has seen the messages and posts on what happened. But because JML completely stepped away from working on Non it's essentially made it so using NewSM is the only option for a lot of people. After all, most of it remains the same. Had he continued to work on the Non projects regardless of the actions of others, merely stating that he disagrees with the changes in NewSM and will continue to maintain NonSM as he sees fit, I truly believe a lot of people would still be using the Non version (I certainly appreciate it's lightness compared to Agordejo and RaySession). 

Of course having one unified standard would be nice but people will always have technical disagreements. Provided all possible standards are open I don't mind this (as oppose to MS's "Embrace, extend, and extinguish" approach) as I can choose which I prefer to use, and if someone doesn't like this then they can maintain a version that works for them (e.g. maybe someone would want a version that plays well with LADISH? Or maybe one with ALSA MIDI ports? There's no chance I'd work on those features, but I have no objection to anyone who implements them in their own fork). 

In the case of NSM clients, it seems being compliant with all three (four?) versions at the same time is a doable thing, so there's little reason for me not to ensure people who prefer other session managers are happy with it too. If that changes then I'll have to choose a standard and stick to it, or start introducing build options/configurations for each depending on my feelings on the variations (I often see the value in multiple approaches to things), but this is not to worry about at the moment anyways.

Free software is all about the freedom to use and change as you desire, I wouldn't want to take that away from others.

Nonetheless, I do understand if you feel different. If you want to reach out privately regarding this you can message me at jamiedrinkell followed by the well known google email domain. (yes, I know gmail is heresy but I've had it for so long that the move to something different is hard!). I'd love to see your work on a this session manager!

**Comment by grammoboy2** [comment]
Thx, but I don't think the potential fun outweighs the potential costs for me atm. It is what it is for now.

**Comment by M0JXD** [comment]
Provided your dependencies would allow it, might you consider a license like the [CC BY-NC-ND?](https://creativecommons.org/licenses/by-nc-nd/4.0/)

This license should allow your code to be used but would prevent people distributing derivative works without your permission. I'm not saying it's a perfect solution, but it might align with what you require.

**Comment by grammoboy2** [comment]
I did consider something like that. Source open, but no license (maybe for a set period). But it's tricky and it would just a matter of time before the shit hits the fan. People just won't accept that and it's me against the world on this topic. Also it might be more fun to have just a open project, which people can enjoy like back in the days. I don't think CC license is for source code, is it? I just didn't figured out a satisfactory solution, so I will take a step back and re-evaluate again later, as this question bugs me from time to time of course. I won't promise anything atm, raysession will suit most users preferences on Cinnamon like environments anyway I think.

**Comment by M0JXD** [comment]
Yeah, it's a toughie. I do agree most people do not seem to have care or respect for licenses anyways. Just look at the forks of [Unity](https://github.com/Unity-Technologies/UnityCsReference), every fork technically counts as a redistribution as several of them have made modifications, all of which are not allowed.

CC isn't ideal/recommended for source code but it can be used. There's no common code specific license that blocks derivative works it seems.

I may prefer more "conventional" desktop paradigms but I'd still like to see your project :)
As said, I do appreciate different approaches!

**Comment by M0JXD** [comment]
I made a new issue for NSM stuff as this discussion is getting quite long for a closed issue. 🙂

**Comment by grammoboy2** [comment]
You don't have for me. I've said what I wanted to say and it became clear to me during this discussion that it's not ready and that I'm not ready and I can't even promise that point will ever come. I'm working on this because I still find it a intriguing subject. I participated in this discussion to get a clearer view on optional-gui capability and how wmctrl can be used, with it's pros an cons, and to share some things I've learned or realized along the way, which may be helpful for others when thinking about these matters.

**Comment by M0JXD** [comment]
Ensuring NSM works is important to me so the issue will remain, there are definite issues at the minute regarding saving the visibility for New and Non SM so I'll fix that anyways, even if I don't like it as a requirement. 

I'll add a config option to disable this too (I think I'll make a file called nsm_config). I think runtime options are better for these aspects as it means the packagers don't have to decide between several different build options, and end users of system packages can still configure as they wish. Given there's not really any additional overhead to support it, just a couple more checks at startup, I think it's better.

If you ever decide you want to release/share your work please give me a ping :)
Alternatively if you want to privately send through a version for me to use as a test bed for making sure Non-Sequencer-Fork works when optional-gui isn't reported by the server, I'll happily do that too :)

**Comment by grammoboy2** [comment]
I've your e-mail, but don't expect anything soon. You're interesting because you're relatively new to NSM, maybe you see things more experienced users don't see. 

Apparently you prefer Raysession above Agordejo, why is that if you could name one thing or maybe two? (Just interested, my mockup doesn't fall into this category. )

By the way, i don't use sway myself. I feel that is in potential maybe the most ideal WM for this from the ones I know, but I prefer the simplicity of dwm / dwl myself, with the 'deck' patch. Also you've to make code changes in the config files, which I see as a advantage, because it prevents you from being busy with the configuration all the time. RiverWM gives more flexibility, but is wayland only.

**Comment by M0JXD** [comment]
A few reasons, some arguably superficial:

- The integrated patchbay is great, I really like that I can drag and connect in stereo pairs which is more tricky in QJackCtl, plus dragging an existing connection removes it which is easier for me too. Evidently there are other nice Patchbay apps but given it's part of a RaySession and it has other features I value, it's the one I use.
- Integrated Git version control is epic and has saved my bacon a few times now.
- Save GUI states is really useful, especially given some of my main NSM programs (e.g. Non-XT) do not save their own states.
- It can be themed and integrates with my Cinnamon/GTK system theme. Agordejo _should_ be able to do this too but it's programmed to force the ugly (IMO) Qt fusion style and indigo theme. New session manager's legacy GUI also removes any sense of customisation which is sad.
- RayHack for NSM incompatible clients, e.g. Rosegarden. And conversion utilities for apps that have only got NSM support more recently. (Haven't needed them but nice to know they're there!)
- The message pane is useful, especially when I need to export the NSM_URL for testing dev builds that are not installed system wide. Without launching in a terminal, Agordejo will not tell you this.
- It's under more active development (Agordejo's last commit was two years ago), plus the main developer is wonderful, helpful and responsive! (And not part of any controversy...)
- Agordejo being hosted on it's own private git server makes it hard for developers to get involved, plus https://laborejo.org/bugs is just a blank web page. I understand the distaste for Github et al but if you want people to be involved with your "community" project a generally accessible solution like GitHub, GitLab, Codeberg or (god forbid) Sourceforge should be used.

My only dislike with RaySession is how heavy it is -  between the GUI, daemon and patch memory it uses nearly 100MB of memory, which is kinda crazy when Timeline and Mixer together use <20MB. But that's the cost you pay for using Python. Agordejo is ~50MB so it's not really memory conscious either. On lower end systems my choice is the original NonSM, but of course that is unmaintained.

**Comment by Houston4444** [comment]
happy to see all these compliments !

About the memory consumption, it is true that it is higher is RS than with the couple nsmd/non_session_manager or nsmd/nsm_legacy_gui, it have not tested Agordejo for a long time. I spend more time optimizing startup time than memory consumption.  Curiously, I was checking this point when I saw this post, I have not exactly the same results (+50Mb for raysession, 13.5Mb for ray-daemon, 9.5Mb for patchbay daemon). Note that only the private memory is important, especially for patchbay_daemon, its shared memory is mostly this one of JACK (and liblo too). Python is in cause, but no so much, it's hard to have a python executable taking less than 10Mb but after it seems to not grow so much. patchbay_daemon could be written in C instead, it is some work, too much I think for winning 10Mb. Convert ray-daemon to C or C++ would be a very long nightmare for me, for the GUI it is the same x 5 ;). There is probably something to do for the GUI, I will inspect.

**Comment by M0JXD** [comment]
It's not really a problem - these days it seems Electron apps are considered ["lightweight"](https://learn.microsoft.com/en-us/shows/visual-studio-code/)...

I think most people's computers can handle it. My previous PC, a 2015 i3 8GB can handle it fine. When I say low end I really mean low end, e.g. my Acer Aspire One 522 with 2GB of RAM or Raspberry Pi Zero. Most any modern day GUI app struggles on those anyways!

Here's what the resource usage is on my PC for what I consider a fairly minimal session:

![Raysession Memory](https://github.com/user-attachments/assets/7b898da7-d902-4123-a7b0-c953a153c14d)

By comparison, the Non-XT suite is nothing. But they make a specific effort to be very light so it is quite an unfair comparison. Most of my system utilities are heavier than them! But it does feel a bit funny that the apps I'm using to do the Audio work are a magnitude lighter than the tool that makes them play together nicely.

![non-mem](https://github.com/user-attachments/assets/5091639e-317e-403d-a540-06f384adeff4)

I'm very happy with RaySession, if a bit more memory usage is the cost for the best session manager around, that's no problem by me :)

**Comment by grammoboy2** [comment]
A 'window switcher / chooser' could improve workflow on a traditional desktop I can imagine. `rofi -show window` or something similar. In theory you could even make such a switcher for nsm client windows specific, but I'm not sure how helpful it would be. 

Non-session-manager is actually quite simple to use, is fast, light and still compiles here. Agordejo has the advantage that it can list sessions based on modification date, which could be implemented in non-session-manager as well probably. I'm not a fan of the 'file watcher' mechanism in the background though and it has some blocking OSC behavior, it seems to send multiple times a message to list sessions and to remove a client. Without the filewatcher system and probably without the option to rename, copy, delete folders (designed in this way), it would have been a better application for me tbh. 

No recent commits shouldn't be seen as a disadvantage imho, that could mean the application is finished in the eyes of the developer and you've more time to spend on fixing session management related issues in JACK or NSM clients. 

EUPL looks like a interesting license.

**Comment by grammoboy2** [comment]
> EUPL looks like a interesting license.

It's compatible with the GPL. EUPL is a reciprocity license, but the GPL not that much ... It has to been seen if the original author has the right to use code changes to their own code, if someone finds a valid reason to use EUPL code in a GPL project. Note that the FSF suggests a illegal route to re-license EUPL to GPL+.  You're not allowed to re-license without a valid reason.  
https://interoperable-europe.ec.europa.eu/collection/eupl/discussion/how-does-fsf-considers-eupl 

Also the nsmd fork changed the license from GPL2+ to GPL3+, this is a right which the GPL2+ provides, but I think this means that JML can't use the code changes in his original GPL2+ version of nsmd, but needs to relicense to GPL3. Meh.  

A hide all button will work. Currently I'm not in favor of a 'hide all and close' functionality. If someone makes a mistake, by forgetting to hide all clients before close, it's easy to do it after opening a session I think.

**Comment by M0JXD** [comment]
Sorry for my delay in response.

EUPL looks good - but it's completely up to you for your project. For me my preferred license is actually MIT when I can use it. Gives people the most freedom to do what they want whilst still maintaining credit to those who did the work. If it was a viable desktop platform I'd be using FreeBSD as my daily! Obviously I can't re-license Non-Sequencer to something permissive but I see no reason to restrict it more with GPL3.

I do like the simplicity of the original NonSM, although having a list of NSM clients to click is nicer than typing executable names. Maybe that wouldn't irk me as much if it had tab-completion.

A hide all would actually be pretty useful to me just as a workflow tool.

One thing I realised is the reason I keep using "Windows" style desktops is because I have to use Windows for work. Learning a whole workflow with a different desktop doesn't benefit me outside of using it at home. But my familiarity with the likes of Cinnamon/KDE/XFCE translates well to when I'm forced to use what is sadly the world's most popular desktop OS. 

And of course the opposite applies. I've been familiar with Windows since I was a kid when XP was king. Desktop environments that mimic Windows made my adaptation to Linux much easier, despite the fact they're radically different underneath.

All to say, supporting such desktops is a good idea if you want to make a tool for all to use. Of course there's nothing wrong with making a tool tailored to the desktop experience you like though. Would be a boring world if we all liked things the same way!

**Comment by grammoboy2** [comment]
> The Zig build system requires you write Zig... I don't know it and I don't fancy trying to use the language intended to be a C replacement for my build system, it's the wrong tool IMO. It will also always be developed with a Zig first mindset. If it suffers an edge case/break for C/C++ building are we sure that will be fixed in a timely manner?

> As far as I see it, Lua is great for this because it's simple but still really fast, unlike Python based ones like Waf. Xmake even beats ninja in some tests when used as a direct building tool. 

(From a other issue) 
Lua is also a language. The advantage of Zig build could be to make a static linked binary, ntk included and might even make crossplatform support within reach. 
https://github.com/allyourcodebase 

> although having a list of NSM clients to click is nicer than typing executable names. 

This can easily fixed. In fact I was messing with it the other day, removed the 'add client' button, so you're forced to launch the NSM client from a other tool, unless there are more then one daemons attached, then the add button appears. 

> All to say, supporting such desktops is a good idea if you want to make a tool for all to use.  

My tools are made for myself indeed, not with the average desktop user in mind, but the whole wmctrl thing made me look at agordejo and non-session-manager, from which the latter I still prefer because of the straightforward and 'direct' design. It was explicitly not the plan to do anything other then for my own workflow ... because ... , but I can't resist apparently , I'm regularly on lxde now ...  let's see how my setup works with non-session-manager and if I can make it all a bit more accessible for others, just in case.

**Comment by M0JXD** [comment]
> (From a other issue)
Lua is also a language.

Lua is an easy to use, popular dynamic scripting language. It has 21 keywords and is arguably easier to learn and use than Python. It's definitely easier than Zig. I've spent less than 20hrs total coding in Lua and have already used it for a build system and adjusting one of my editor's modules to my needs, that's how easy it is.

Maybe I was a bit quick to dismiss using Zig's build system, but I'm trying to make my time learning/programming count. While programming is only an occasional task of my day job, I need to follow my industry's (embedded) trends to progress. At the moment, the technologies I need to be best at right now are C (with some C++) and Python. Lua is also popular in the embedded sector due to how it easily integrates with C environments, and I've found I prefer it to Python too. Long term, the future is clearly Rust (Rust and ADA are the only suitable languages on the [NSA list](https://readwrite.com/the-nsa-list-of-memory-safe-programming-languages-has-been-updated/) that can target bare-metal). So my focus is on these four languages. On top of that there's things like DSA, DSP and a wealth of other stuff I need to learn and get better at!

Zig does not chart the top 50 programming languages on TIOBE. The job market is practically non-existent. I don't even use any software made with Zig - the only really popular project I've heard of is Ghostty. Most of the stuff I'd need to work with in Zig to do hobby projects I like, such as GUI, graphics and audio libraries, are all in alpha/unfinished states. (I'm finding this to be the case in Rust a bit as well.)

> The advantage of Zig build could be to make a static linked binary, ntk included and might even make crossplatform support within reach.
https://github.com/allyourcodebase

Xmake can do all this. Actually, cross-compilation is one of it's strengths. It's built in package manager is also really cool. CMake (which is sadly the industry standard) also works for this - Stazed has already set up statically linked build options for Non-XT, they're actually needed for Non-Mixer. I really dislike Cmake's DSL though so didn't use it.

However having the capabilities in the build system is only half the issue. I had a go at building Non-Timeline-XT on Windows a little while back, but the Non projects have a heavy reliance on GNU extensions like asprintf() and the POSIX API, e.g. posix_memalign(). These are not implemented in MinGW and porting all these aspects would've been quite the task. Without a session manager that supports other OSes, I asked myself what the point would be? [As you clearly saw there's not even a guarantee that NSM works on Windows anyways.](https://github.com/Houston4444/RaySession/issues/111) I refuse to own any Apple devices, and I'm not going to break their EULA's, so I can't do anything for macOS either.

I just can't reason with myself to learn Zig just to use it's build system in a hobby project, when the system I'm already using is very adequate. If a Zig zealot want's to contribute it as an alternative build system I'm open to forks or PR's, although I'd ask that they check in for maintenance of a Zig build now and then before merging.

I want to add here, if there's a movement to make the software of Linux Audio more portable I'm certainly up for the challenge of porting the Non projects to new systems. I've seen even Haiku is getting a lot of stuff ported to it these days.

**Comment by M0JXD** [comment]
Just to add, if you want static builds and more platform support please open the issues for them :)

**Comment by M0JXD** [comment]
> My tools are made for myself indeed, not with the average desktop user in mind, but the whole wmctrl thing made me look at agordejo and non-session-manager, from which the latter I still prefer because of the straightforward and 'direct' design. It was explicitly not the plan to do anything other then for my own workflow ... because ... , but I can't resist apparently , I'm regularly on lxde now ... let's see how my setup works with non-session-manager and if I can make it all a bit more accessible for others, just in case.

I like LXDE too (more so than LXQt) and I'm glad some new maintainers have stepped up. If you do share your session manager and it has stuff for LXDE I'm sure I could make use of those aspects in Cinnamon.

**Comment by grammoboy2** [comment]
FYI: One thing I just discovered recently: 

In WM's like dwm and river wm (I like this one for wayland), you can toggle each workspace or tag to be visible. 

So if you've non-sequencer on workspace 1, non-mixer on 2 and non-timeline on 3 and workspace 1 is focused. You can toggle non-mixer and / or non-timeline visibility on workspace 1 with a keybinding (Super+Ctrl+TagNum). 

In the WM configuration you can set non-sequencer to always start on workspace x etc.

Could be a convenient workflow for non-daw.

**Comment by M0JXD** [comment]
Hey, thanks for sharing, that sounds really useful. I use a few different workflows with Non and using workspaces is a great one, but this visibility switching would be pretty handy too.

I must get back on to this at some point, the NSM C API branch seems quite functional from my use so I'll merge that to the master next chance I have. As said I only intended to do minimal maintenance, but I do want to improve the NSM compliance as per #5 .

I have kinda been struggling with the lack of a more conventional "MIDI track" in Non as you mentioned in #2 . Qtractor works nice as an alternative and syncs with JACK-Transport. But I'm also struggling using such unconventional DAW choices with little cross-platform and community support. I keep using REAPER on and off because I run into limitations/difficulties, might just cave, get the license and use it as my main DAW.

I still like Non and it's nice to use on projects that aren't too complex, or as an option on resource constrained hardware. I think it's worth keeping alive as a super light music making option as nothing else comes close to the lightness, closed or open source. This would make it worth reviving the NTK version of NSM and making it NewSM compliant.

FYI at some point I want to move my open source repos to a new platform (likely Codeberg). Ofc I'll keep a GitHub account (and can possibly set up mirrors) so I can contribute to other projects, but it will no longer be "home" for anything I'm maintaining.

---

Edit 08/02/26: My projects will stay on GitHub. At the time I was having a bit of a personal thing around copyright law and wanted to get away from it all. I'm still not happy with how everything is, but it's undeniable that GitHub continues to be the best and most popular place to host code and offers many useful extras.

**Comment by grammoboy2** [comment]
> * It's under more active development (Agordejo's last commit was two years ago)

Project archived: 

https://www.laborejo.org/agordejo/ 
https://www.laborejo.org

My goodness.

**Comment by M0JXD** [comment]
> Project archived:
> 
> https://www.laborejo.org/agordejo/ https://www.laborejo.org
> 
> My goodness.

Wow, that is a surprise.
That basically means RaySession is it for supported session managers. New Session Manager has not undergone much maintenance either, and given Nils is one of the maintainers I'm not sure whether that is effected too? 

Also a shame because Patroneo was a nice program that made MIDI sequencing fun. In my experience Python programs get outdated rather quick so I wonder how long these will be usable without a maintainer.

**Comment by M0JXD** [comment]
Hi again all those who where dragged into this VERY long issue thread last year.

As [I mentioned earlier](https://github.com/M0JXD/non-sequencer-fork/issues/3#issuecomment-2608011751), I'm now on a bit of mission to make FreeBSD a viable reality for my desktop computing needs. One thing it lacks is a good musician's audio recording system so I've started work trying to [port the Non-XT series](https://github.com/Stazed/non-mixer-xt/pull/61).

The biggest issue is [the lack of a session manager](https://github.com/Stazed/non-mixer-xt/issues/58#issuecomment-3862244554) for NSM enabled applications (several like Qtractor, Hydrogen and Guitarix are already ported to FreeBSD). This is going to be my first task when I get round to it.

My temptation is to fork the original NonSM. I would like to bring it up to date with optional NewSM compliance, NTK/FLTK builds as Non-XT does and some other niceties. However @grammoboy2 you mentioned you were making your own session manager. I'm wondering how that is going and whether it is perhaps a better option for a FreeBSD port? Porting Raysession is also an option but might be harder as it bundles many extra features beyond session management which would need vetted before saying the port is good.

As for Non-Sequencer - this will be my last priority. At the end of the day, I'm an instrumentalist and I don't really use MIDI often, usually all I need is a drum track which Hydrogen is a better option for. Hence why I've done basically nothing more on it, but still I would like to keep it working as it'd be nice to have the Non series continue to live. We've already lost too many Linux audio apps like [Wired](https://sourceforge.net/projects/wired/) and [Jokosher](https://github.com/mjumbewu/jokosher).

**Comment by grammoboy2** [comment]
Didn't looked at the code for a year, but at least it's in my mind again. Still plans to release at some point, but nothing concrete yet.

**Comment by M0JXD** [comment]
Thanks for the update :) hope you do choose to share it at some point.

**Comment by grammoboy2** [comment]
Thinking about AGPL for the daemon and GPL 3 for the tools

**Comment by M0JXD** [comment]
> Thinking about AGPL for the daemon and GPL 3 for the tools

Hey again, sorry for the delay! That sound's good, it'll be great to see your implementation :)
